### PR TITLE
Simva

### PR DESCRIPTION
**Pull request by Victorma**

### CODE REVIEW

File: Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+            {
+                yield return new Tuple<string, string>(www.downloadHandler.text, null);
+            }else
+            {
+                yield return new Tuple<string, string>(null, www.downloadHandler.text);
+            }
+        }
+    }
+
+    public IEnumerator getSchedule()
```
Comment (Victorma) [inline Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs]: Caps

File: Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+            }
+            else
+            {
+                this.schedule = JSON.Parse(www.downloadHandler.text);
+                yield return new Tuple<string, string>(null, www.downloadHandler.text);
+            }
+        }
+    }
+
+    public IEnumerator getTarget(string activityId)
```
Comment (Victorma) [inline Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs]: Caps

File: Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+            if (www.responseCode != 200)
+            {
+                yield return new Tuple<string, string>(www.downloadHandler.text, null);
+            }
+            else
+            {
+                yield return new Tuple<string, string>(null, www.downloadHandler.text);
+            }
+        }
+    }
```
Comment (Victorma) [inline Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs]: All these code seems to be duplicated following the same structure and request handling. I suggest using another coroutine for response handling

File: Assets/uAdventure/__Scripts/Runner/GameLogic/GUIManager.cs [hunk 4]
```diff
@@ -373,53 +375,72 @@ public void ClearData()
...
+                }
+                else
+                {
+                    StartCoroutine(Continue(SimvaController.Instance.getCurrentActivityId(), true));
+                }
+            }
+            else
+            {
+                Application.Quit();
+            }
```
Comment (Victorma) [inline Assets/uAdventure/__Scripts/Runner/GameLogic/GUIManager.cs]: I still think this code should belong to analyticsextension.cs and should be plugged to the game.cs lifecycle any form

File: Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+            loadingManager.IsLoading(state);
+        }
+    }
+
+    public void InitUser()
+    {
+        this.StartCoroutine(LoginAndSchedule());
+    }
+
+    public string getCurrentActivityId()
```
[the remark this answers is not in the document]
Comment (Victorma) [reply Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs]: both fixed

File: Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+        }
+    }
+
+    public void NotifyLoading(bool state)
+    {
+        foreach (SimvaLoadingManager loadingManager in loadingManagers)
+        {
+            loadingManager.IsLoading(state);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Victorma) [reply Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs]: Fixed

File: Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+
+    public IEnumerator LoginAndSchedule()
+    {
+        NotifyLoading(true);
+        CoroutineWithData cd = new CoroutineWithData(this, Login());
+        yield return cd.coroutine;
+
+        Tuple<string, string> result = (Tuple<string, string>) cd.result;
```
[the remark this answers is not in the document]
Comment (Victorma) [reply Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs]: Cannot be done

File: Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+            return schedule["next"].Value;
+        }
+        return null;
+    }
+
+    public JSONNode getActivity(string activityId)
+    {
+        if (schedule != null)
+        {
+            return schedule["activities"][activityId];
```
[the remark this answers is not in the document]
Comment (Victorma) [reply Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs]: JSON keys cannot be easily checked

File: Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs (hunk 3 above)
Comment (Victorma) [reply Assets/uAdventureAnalytics/Plugins/Simva/SimvaController.cs]: Fixed with ProcessWWW method

File: Assets/uAdventure/__Scripts/Runner/GameLogic/GUIManager.cs (hunk 4 above)
Comment (Victorma) [reply Assets/uAdventure/__Scripts/Runner/GameLogic/GUIManager.cs]: Fixed by adding a OnGameFinished method to all extensions preventing the game to be closed and hooking it

**Comment by Synpheros** [comment]
Then, how shall we proceed to merge this branch? What, improvements should be made by me?

**Pull request merged**